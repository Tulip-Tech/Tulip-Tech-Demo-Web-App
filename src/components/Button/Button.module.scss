@use 'src/styles/variables';
@use 'src/styles/theme';
@use 'src/styles/mixins/responsive';

$small-button-height: 28px;
$medium-button-height: 36px;
$large-button-height: 40px;

.button {
  position: relative;
  display: inline-flex;
  justify-content: inherit;
  align-items: center;
  height: $medium-button-height;
  padding: 0 16px;
  overflow: hidden;

  font-family: var(--body-alt-font-family);
  font-weight: var(--body-font-weight-bold);
  font-size: 16px;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  border-radius: 4px;
  outline: none;
  cursor: pointer;
  transition: background-color 0.1s ease, transform 0.1s ease;

  @media (hover: hover) and (pointer: fine) {
    &:not(.disabled) {
      &:hover,
      &:focus {
        z-index: 1;
        transform: scale(1.1);
      }

      &:focus:not(:focus-visible):not(:hover) {
        transform: scale(1);
      }

      &:focus-visible {
        transform: scale(1.1);
      }
    }
  }

  &.disabled {
    cursor: default;
    opacity: 0.5;
  }

  &.small {
    height: $small-button-height;
  }

  &.large {
    height: $large-button-height;
  }

  &.default {
    color: currentColor;
    background-color: theme.$btn-default-bg;
  }

  &.primary {
    color: var(--highlight-contrast-color, theme.$btn-primary-color);
    background-color: var(--highlight-color, theme.$btn-primary-bg);
  }

  &.headerbtn {
    width: max-content;
    padding: 5px 5px 5px 8px;

    color: variables.$white;
    font-weight: 800;
    font-size: 14px;
    background-color: var(--highlight-color, theme.$btn-primary-bg);
    border-radius: 1px;
  }
  &.outlined {
    width: auto;
    border: 1px solid rgba(255, 255, 255, 0.3);

    &:not(.disabled) {
      &.active,
      &:focus {
        color: var(--highlight-contrast-color, theme.$btn-primary-color);
        background-color: var(--highlight-color, theme.$btn-primary-bg);
        border-color: var(--highlight-color, theme.$btn-primary-bg);
      }
    }
  }

  &.text {
    background: none;
    opacity: 0.7;

    &:not(.disabled) {
      &.active,
      &:focus {
        opacity: 1;
      }

      &:hover {
        z-index: 1;
        background: theme.$btn-default-bg;
        opacity: 1;
      }
    }
  }

  &.fullWidth {
    justify-content: center;
    width: 100%;

    @media (hover: hover) and (pointer: fine) {
      &:hover:not(.disabled) {
        transform: scale(1.04);
      }
    }
  }

  > svg {
    width: 18px;
    height: 18px;
  }
}

.startIcon {
  display: flex;
  align-items: center;
  height: 100%;
  margin-right: 11px;

  > svg {
    width: 20px;
    height: 20px;
    fill: currentColor;

    @include responsive.tablet-only() {
      width: 18px;
      height: 18px;
    }
  }
}

.hidden {
  visibility: hidden;
}

.centerAbsolute {
  position: absolute;
  right: 0;
  left: 0;
  margin: auto;
  transform: translate(-5px, -5px);
}
