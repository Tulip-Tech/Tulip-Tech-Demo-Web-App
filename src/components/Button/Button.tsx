import React, { MouseEventHandler } from 'react';
import classNames from 'classnames';
import { NavLink } from 'react-router-dom';

import styles from './Button.module.scss';

import Spinner from '#components/Spinner/Spinner';

type Color = 'default' | 'primary' | 'headerbtn';

type Variant = 'contained' | 'outlined' | 'text';

type Props = {
  children?: React.ReactNode;
  label: string;
  active?: boolean;
  color?: Color;
  fullWidth?: boolean;
  startIcon?: JSX.Element;
  variant?: Variant;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  tabIndex?: number;
  size?: 'small' | 'medium' | 'large';
  to?: string;
  role?: string;
  className?: string;
  type?: 'button' | 'submit' | 'reset';
  disabled?: boolean;
  busy?: boolean;
  id?: string;
} & React.AriaAttributes;

const Button: React.FC<Props> = ({
  label,
  children,
  color = 'default',
  startIcon,
  fullWidth = false,
  active = false,
  variant = 'outlined',
  size = 'medium',
  disabled,
  busy,
  type,
  to,
  onClick,
  className,
  ...rest
}: Props) => {
  const buttonClassName = (isActive: boolean) =>
    classNames(styles.button, className, styles[color], styles[variant], {
      [styles.active]: isActive,
      [styles.fullWidth]: fullWidth,
      [styles.large]: size === 'large',
      [styles.small]: size === 'small',
      [styles.disabled]: disabled,
    });

  const content = (
    <>
      {startIcon && <div className={styles.startIcon}>{startIcon}</div>}
      {<span className={classNames(styles.buttonLabel, { [styles.hidden]: busy }) || undefined}>{label}</span>}
      {children}
      {busy && <Spinner className={styles.centerAbsolute} size={'small'} />}
    </>
  );

  if (to) {
    return (
      <NavLink className={({ isActive }) => buttonClassName(isActive)} to={to} {...rest} end>
        {content}
      </NavLink>
    );
  }

  return (
    <button className={buttonClassName(active)} onClick={onClick} type={type} disabled={disabled} aria-disabled={disabled} {...rest}>
      {content}
    </button>
  );
};
export default Button;
