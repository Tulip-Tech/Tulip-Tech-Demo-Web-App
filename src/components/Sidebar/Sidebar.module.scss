@use 'src/styles/variables';
@use 'src/styles/theme';
@use 'src/styles/mixins/responsive';

//
// jwSidebarBackdrop
// --------------------------------

.backdrop {
  position: fixed;
  top: 0;
  left: 0;
  z-index: variables.$sidebar-z-index - 1;
  display: none;
  width: 100%;
  height: 100%;
  background: rgba(variables.$black, 0.4);
  transition: all 0.3s ease;
}

//
// jwSidebar
// --------------------------------

.sidebar {
  position: fixed;
  top: 0;
  z-index: variables.$sidebar-z-index;
  display: none;
  width: 270px;
  max-width: 90vw;
  height: 100vh;
  background-color: var(--body-background-color);
  transform: translateX(-100%);
  transition: transform 0.3s cubic-bezier(0.52, 0.51, 0.2, 1);
}

.heading {
  display: flex;
  padding: variables.$base-spacing 0 0;
}

.group {
  display: flex;
  flex-direction: column;
  padding: variables.$base-spacing 0;
}

//
// mediaQueries
// --------------------------------

@include responsive.mobile-and-tablet() {
  .sidebar {
    display: inline-block;

    &.open {
      transform: translateX(0);
    }
  }

  .backdrop {
    display: inline-block;
    visibility: hidden;

    &.visible {
      visibility: visible;
    }
  }
}
