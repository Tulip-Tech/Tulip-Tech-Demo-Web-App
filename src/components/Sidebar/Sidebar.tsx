import React, { Fragment, ReactFragment } from 'react';
import classNames from 'classnames';
import { useTranslation } from 'react-i18next';

import styles from './Sidebar.module.scss';

import Close from '#src/icons/Close';
import IconButton from '#components/IconButton/IconButton';

type SidebarProps = {
  isOpen: boolean;
  onClose: () => void;
  children?: ReactFragment;
};

const Sidebar: React.FC<SidebarProps> = ({ isOpen, onClose, children }) => {
  const { t } = useTranslation('menu');

  return (
    <Fragment>
      <div
        className={classNames(styles.backdrop, {
          [styles.visible]: isOpen,
        })}
        onClick={onClose}
      />
      <div
        className={classNames(styles.sidebar, {
          [styles.open]: isOpen,
        })}
      >
        <div className={styles.heading}>
          <IconButton onClick={onClose} aria-label={t('close_menu')} tabIndex={isOpen ? 0 : -1}>
            <Close />
          </IconButton>
        </div>
        <nav className={styles.group} onClick={onClose}>
          {children}
        </nav>
      </div>
    </Fragment>
  );
};

export default Sidebar;
