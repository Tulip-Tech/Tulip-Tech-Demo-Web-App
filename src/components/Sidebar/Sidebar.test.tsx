import React from 'react';

import Sidebar from './Sidebar';

import { renderWithRouter } from '#test/testUtils';
import Button from '#components/Button/Button';

describe('<SideBar />', () => {
  const playlistMenuItems = [<Button key="key" label="Home" to="/" />];

  test('renders sideBar', () => {
    const { container } = renderWithRouter(
      <Sidebar isOpen={true} onClose={vi.fn()}>
        {playlistMenuItems}
      </Sidebar>,
    );

    expect(container).toMatchSnapshot();
  });
});
