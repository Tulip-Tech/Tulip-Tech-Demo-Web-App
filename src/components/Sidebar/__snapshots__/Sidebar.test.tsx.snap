// Vitest Snapshot v1

exports[`<SideBar /> > renders sideBar 1`] = `
<div>
  <div
    class="_backdrop_577f70 _visible_577f70"
  />
  <div
    class="_sidebar_577f70 _open_577f70"
  >
    <div
      class="_heading_577f70"
    >
      <div
        aria-label="close_menu"
        class="_iconButton_0fef65"
        role="button"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="_icon_1e9999"
          focusable="false"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </g>
        </svg>
      </div>
    </div>
    <nav
      class="_group_577f70"
    >
      <a
        aria-current="page"
        class="_button_f8f296 _default_f8f296 _outlined_f8f296 _active_f8f296"
        href="/"
      >
        <span>
          Home
        </span>
      </a>
    </nav>
  </div>
</div>
`;
